cobalt/base/process/process_metrics_helper.cc

```diff
@@ -0,0 +1,203 @@
+}
+
+}  // namespace
+
+// static
+int ProcessMetricsHelper::GetClockTicksPerS() {
+  return clock_ticks_per_s.load();
+}
+
+// static
+int ProcessMetricsHelper::GetClockTicksPerS(ReadCallback uptime_callback,
+                                            ReadCallback stat_callback) {
+  double uptime = 0.0;
+  {
+    auto uptime_contents = std::move(uptime_callback).Run();
+    if (!uptime_contents) return 0;
+    auto parts = SplitString(*uptime_contents, " ", TRIM_WHITESPACE,
+                             SPLIT_WANT_NONEMPTY);
+    if (parts.size() == 0 || !StringToDouble(parts[0], &uptime) ||
+        uptime == 0.0) {
+      return 0;
+    }
+  }
+
+  // Get starttime.
+  auto fields = GetProcStatFields(std::move(stat_callback), {21});
+  if (fields.size() != 1) return 0;
+  double starttime;
+  if (!StringToDouble(fields[0], &starttime) || starttime == 0.0) return 0;
+  int ticks_per_s = static_cast<int>(starttime / uptime);
+  int nearest_tens = 10 * ((ticks_per_s + 5) / 10);
+  return nearest_tens;
+}
+
+// static
+void ProcessMetricsHelper::PopulateClockTicksPerS() {
+  DCHECK_EQ(clock_ticks_per_s.load(), 0);
+  clock_ticks_per_s.store(
+      GetClockTicksPerS(GetReadCallback(FilePath("/proc/uptime")),
+                        GetReadCallback(FilePath("/proc/self/stat"))));
+}
+
+// static
+PlatformThreadId ProcessMetricsHelper::GetPid() {
+  PlatformThreadId pid;
+  auto fields = GetProcStatFields(FilePath("/proc/self"), {0});
+  if (fields.size() != 1 || !StringToInt(fields[0], &pid)) {
+    return -1;
+  }
+  return pid;
+}
+
+// static
+TimeDelta ProcessMetricsHelper::GetCumulativeCPUUsage() {
+  int ticks_per_s = clock_ticks_per_s.load();
+  if (ticks_per_s == 0) return TimeDelta();
+  return GetCPUUsage(FilePath("/proc/self"), ticks_per_s);
+}
+
+// static
+base::Value ProcessMetricsHelper::GetCumulativeCPUUsagePerThread() {
+  int ticks_per_s = clock_ticks_per_s.load();
+  if (ticks_per_s == 0) return base::Value();
+  base::Value::List cpu_per_thread;
+  FileEnumerator file_enum(FilePath("/proc/self/task"), /*recursive=*/false,
+                           FileEnumerator::DIRECTORIES);
+  for (FilePath path = file_enum.Next(); !path.empty();
+       path = file_enum.Next()) {
+    // tid, thread name, utime, and stime.
+    Fields fields = GetProcStatFields(path, {0, 1, 13, 14});
+    if (fields.size() != 4) continue;
+    int id;
+    if (!StringToInt(fields[0], &id)) continue;
+    base::Value::Dict entry =
+        base::Value::Dict()
+            .Set("id", id)
+            .Set("name", fields[1])
+            .Set("utime", fields[2])
+            .Set("stime", fields[3])
+            .Set("usage_seconds",
+                 CalculateCPUUsageSeconds(fields[2], fields[3], ticks_per_s));
+    cpu_per_thread.Append(std::move(entry));
+  }
+  return base::Value(std::move(cpu_per_thread));
+}
+
+// static
+ProcessMetricsHelper::Fields ProcessMetricsHelper::GetProcStatFields(
```

Comment:
I mentioned this before, but this basically ends being a re-write of Linux `ParseProcStats` in https://github.com/youtube/cobalt/blob/main/base/process/internal_linux.cc#L78

Ideally, we could avoid the re-write and simply call into the existing function, e.g. include `internal_linux.h` here. If this gets too hacky though and requires too much patching of the original file, its okay as long as we are super defensive here and return on any unexpected formats.